### PR TITLE
feat: allow to specify preload fetch priority

### PR DESCRIPTION
**Pull request by dargmuesli**
I'd like to be able to set the [fetch priority](https://web.dev/fetch-priority/) for preloaded images. This is a quick draft of an implementation. Happy to hear feedback :heart:

### CODE REVIEW

**Comment by Baroshem** [comment]
@dargmuesli What would be the reason to have this functionality?

If you set preload it would most probably mean that you would like the image to have a high priority IMO. 

Setting it to auto may result in some issues of having a preload and fetchPriority low.

If I get this completely wrong, please let me know. Maybe I am missing something :)

**Comment by dargmuesli** [comment]
Of course the main goal is to be able to specify "high" fetch priority for a preloaded image. I just included "low" and "auto" as it's part of the specification.

File: src/runtime/components/nuxt-picture.ts [hunk 1]
```diff
@@ -54,7 +54,14 @@ export default defineComponent({
     const lastSourceIndex = computed(() => sources.value.length - 1)
 
     if (props.preload) {
-      const link: any = { rel: 'preload', as: 'image', imagesrcset: sources.value[0].srcset }
+      const link: any = {
+        rel: 'preload',
+        as: 'image',
+        imagesrcset: sources.value[0].srcset,
+        ...(typeof props.preload !== 'boolean' && props.preload.fetchPriority
+          ? { fetchpriority: props.preload.fetchPriority }
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/runtime/components/nuxt-picture.ts:62]: [MDN speaks about `fetchPriority`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/fetchPriority) but links to the [HTML's standard `fetchpriority`](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-fetchpriority). I just chose to follow the standard here and use the camelCase version in the JS context here.

**Comment by Baroshem** [comment]
@dargmuesli Isnt it set already by default? I thought it is set like that when passing preload. 

I was looking recently at the unpic library where this is done automatically so maybe I assumed that it is done here as well: If it doesnt work like that, please ignore my previous comment and proceed with thw implementation.

Just review the question from Daniel and you should be good to go 😉

File: src/runtime/components/nuxt-picture.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Baroshem) [reply src/runtime/components/nuxt-picture.ts:62]: Wouldnt it be better to stick to one convention everywhere to avoid typos?

**Review by Baroshem — COMMENTED** [review]
Some comments from my side

**Comment by dargmuesli** [comment]
I tried it in Firefox. Using preload only made let the request have low priority afair, setting it to high made let it be of high priority.

File: src/runtime/components/nuxt-picture.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/runtime/components/nuxt-picture.ts:62]: I would stick with the html specification for what lands in html and with the js convention for what targets js.
What's debatable is whether we follow the mdn wording or the html's specification for which casing lands in html, I'd say, and I'd stick to the html specification there.

**Comment by Javid-Izadfar** [comment]
any resolution for this PR?

File: src/runtime/components/nuxt-picture.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/runtime/components/nuxt-picture.ts:64]: Now, `zhead` types ask for `fetchpriority` to be lowercase only. I think this can be resolved.

**Comment by dargmuesli** [comment]
I resolved the merge conflicts. Reviews welcome!

**Comment by dargmuesli** [comment]
I'm not sure, but to me it looks like the netlify image cdn snapshot need to be updated. If that's incorrect, I'd need assistance to understand how to fix the pipeline instead :pray:

File: src/runtime/components/nuxt-img.ts [hunk 2]
```diff
@@ -94,6 +94,9 @@ export default defineComponent({
                 imagesizes: sizes.value.sizes,
                 imagesrcset: sizes.value.srcset,
               }),
+          ...(typeof props.preload !== 'boolean' && props.preload.fetchPriority
+            ? { fetchpriority: props.preload.fetchPriority }
+            : {}),
```
Comment (danielroe) [inline src/runtime/components/nuxt-img.ts:99]: should we default to `fetchpriority: 'high'`?

**Review by danielroe — APPROVED** [review]

File: src/runtime/components/nuxt-img.ts (hunk 2 above)
Comment (dargmuesli) [reply src/runtime/components/nuxt-img.ts:101]: Sorry for not answering in time, this comment fell off my radar 🌠

Comment (danielroe) [reply src/runtime/components/nuxt-img.ts:101]: Still time to update - what do you think?

Comment (dargmuesli) [reply src/runtime/components/nuxt-img.ts:101]: Why should the fetch priority default to high in your opinion? I'd say we can just leave it for the users to decide if they want to specify any.

**Comment by ausir0726** [comment]
@dargmuesli 
Thank you for the epic contribution, I came here early following the update log.
I recently encountered an issue with download priority.
Although the popularity of fetchPriority is not high, it has already solved most problems based on the UA currently observed.

Also, I would like to ask why fetchPriority needs to be placed separately in the preload object?
Because writing it this way makes the code relatively difficult to maintain,
or does preload have other expected extended functions?

According to my wish, the first setting of the carousel should be set to high as an example
```
// v-for
:preload="{ fetchPriority: index == 0 ? 'high' : 'low'}"
```
If simply using fetchPriority as an attribute
```
// v-for
:fetchPriority="index == 0 ? 'high' : 'low'"
```

**Comment by dargmuesli** [comment]
I think I implemented it that way because Nuxt image includes preload link addition to the document's head. Having done a quick research, I think that it's possible to set the `fetchPriority` on the image element only. So a PR adding `fetchPriority` as an additional property might be a good idea if that's true. But it wouldn't replace the current implementation. It would only correct it so that the preload's `fetchPriority` would be set on the preload link element only.

**Comment by ausir0726** [comment]
@dargmuesli 
That's a great idea. 
Additionally, I noticed that you mentioned fetchPriority in your writing, which could possibly be a boolean. 
Perhaps when the boolean is true, it can be set to high, and when false, it can remain unchanged. 
If fetchPriority is a string, it can provide the user's desired operation, achieving this.

```
// v-for
:fetchPriority="index == 0"
// Expecting to fetchpriority="high"
// else do nothing ( not low or auto ) , do not add any links to the head
```

**Comment by dargmuesli** [comment]
Hmm, I'd favor a string as there can be the "enum" values high, low and auto. I see where you're coming from with the true=high suggestion, but that assumes that people understand fetch priority as a tool to increase the priority, but I think people could also want to use it to lower an image's priority.

**Pull request merged**